File: hash/java/src/main/java/io/rosetta/imagehash/PHash.java
package io.rosetta.imagehash;

import io.rosetta.imagehash.internal.BufferedImageRgb;
import io.rosetta.imagehash.internal.Dct;
import io.rosetta.imagehash.internal.LanczosFixed;
import io.rosetta.imagehash.internal.PilGrayscale;

import java.awt.image.BufferedImage;
import java.util.Arrays;

/**
 * phash: grayscale -> Lanczos to (N*4, N*4) -> 2-D DCT-II -> top-left NxN block
 * -> bit = (coefficient &gt; median + {@link #SNAP_EPS}).
 *
 * <p>The snap-to-threshold tie-break (&epsilon; = 1e-10) deterministically maps
 * coefficients within &epsilon; of the median to bit 0, eliminating cross-port
 * FP-noise divergence at large hash sizes. See spec/SPEC.md
 * &sect;"Threshold tie-break".
 */
public final class PHash {
    /**
     * &epsilon; threshold for the snap-to-threshold tie-break used by PHash,
     * PHashSimple, WHashDb4, and WHashDb4Robust. Coefficients within
     * SNAP_EPS of the threshold map deterministically to bit 0 across all
     * ports. Fixed across all 6 ports — do not vary per-call without a
     * coordinated spec change.
     */
    public static final double SNAP_EPS = 1e-10;

    public static ImageHash compute(BufferedImage img, int hashSize) {
        return compute(img, hashSize, 4);
    }

    public static ImageHash compute(BufferedImage img, int hashSize, int highfreqFactor) {
        if (hashSize < 2)
            throw new IllegalArgumentException("hashSize must be >= 2");

        int imgSize = hashSize * highfreqFactor;

        int[][] rgb = BufferedImageRgb.toIntArray(img);
        int h = rgb.length;
        int w = rgb[0].length;
        int[][] gray = new int[h][w];
        for (int y = 0; y < h; y++) {
            for (int x = 0; x < w; x++) {
                int p = rgb[y][x];
                gray[y][x] = PilGrayscale.toGray((p >> 16) & 0xFF, (p >> 8) & 0xFF, p & 0xFF);
            }
        }

        int[][] resized = LanczosFixed.resize(gray, imgSize, imgSize);
        double[][] doubles = new double[imgSize][imgSize];
        for (int y = 0; y < imgSize; y++)
            for (int x = 0; x < imgSize; x++)
                doubles[y][x] = resized[y][x];

        double[][] dct = Dct.dct2d(doubles);

        double[] block = new double[hashSize * hashSize];
        int k = 0;
        for (int y = 0; y < hashSize; y++)
            for (int x = 0; x < hashSize; x++)
                block[k++] = dct[y][x];

        double[] sorted = block.clone();
        Arrays.sort(sorted);
        int n = sorted.length;
        double median = (n % 2 == 1) ? sorted[n / 2] : (sorted[n / 2 - 1] + sorted[n / 2]) / 2.0;

        // Snap-to-threshold tie-break: deterministic bit 0 on ties.
        double threshold = median + SNAP_EPS;
        boolean[][] bits = new boolean[hashSize][hashSize];
        for (int y = 0; y < hashSize; y++)
            for (int x = 0; x < hashSize; x++)
                bits[y][x] = dct[y][x] > threshold;

        return new ImageHash(bits);
    }

    private PHash() {}
}

File: hash/swift/RosettaSquintHash/Sources/RosettaSquintHash/PHash.swift
import Foundation

/// ε threshold for the snap-to-threshold tie-break used by `phash`,
/// `phashSimple`, `whashDb4`, and `whashDb4Robust`. Coefficients within
/// `SNAP_EPS` of the threshold map deterministically to bit 0 across all
/// ports. Fixed across all 6 ports. See spec/SPEC.md §"Threshold tie-break".
public let SNAP_EPS: Double = 1e-10

/// phash: grayscale → Lanczos to (N*F, N*F) → 2-D DCT → top-left NxN block
/// → bit = (coefficient > median + `SNAP_EPS`).
///
/// The snap-to-threshold tie-break (ε = 1e-10) deterministically maps
/// coefficients within ε of the median to bit 0, eliminating cross-port
/// FP-noise divergence at large hash sizes.
public func phash(_ image: RGBImage, hashSize: Int, highfreqFactor: Int = 4) throws -> Hash {
	try image.validate()
	guard hashSize >= 2 else { throw ImageHashError.invalidHashSize(hashSize) }
	let imgSize = hashSize * highfreqFactor

	let rgb = ImgRGB.toRGB(image)
	let gray = rgbToGray(rgb.data, width: rgb.width, height: rgb.height)
	let resized = lanczosResize(gray, srcW: rgb.width, srcH: rgb.height, dstW: imgSize, dstH: imgSize)

	var doubles = [Double](repeating: 0, count: imgSize * imgSize)
	for i in 0..<resized.count { doubles[i] = Double(resized[i]) }

	let dctOut = dct2d(doubles, imgSize)

	// Extract top-left hashSize x hashSize block
	var block = [Double](repeating: 0, count: hashSize * hashSize)
	var k = 0
	for y in 0..<hashSize {
		for x in 0..<hashSize {
			block[k] = dctOut[y * imgSize + x]
			k += 1
		}
	}
	let sorted = block.sorted()
	let n = sorted.count
	let median: Double = n % 2 == 1 ? sorted[(n - 1) / 2] : (sorted[n / 2 - 1] + sorted[n / 2]) / 2.0

	// Snap-to-threshold tie-break: deterministic bit 0 on ties.
	let threshold = median + SNAP_EPS
	var bits: [[Bool]] = []
	for y in 0..<hashSize {
		var row = [Bool](repeating: false, count: hashSize)
		for x in 0..<hashSize {
			row[x] = dctOut[y * imgSize + x] > threshold
		}
		bits.append(row)
	}
	return try Hash(bits: bits)
}
